aip/0163.md

```diff
@@ -0,0 +1,43 @@
+---
+aip:
+  id: 163
+  state: reviewing
+  created: 2019-12-16
+permalink: /163
+redirect_from:
+  - /0163
+---
+
+# Validating changes
+
+Occasionally, a user wants to validate an intended change to see what the
+result will be before actually making the change. For example, a request to
+provision new servers in a fleet will have an impact on the overall fleet size
+and cost, and could potentially have unexpected downstream effects.
+
+## Guidance
+
+APIs **may** provide an option to validate, but not actually execute, a
+request, and provide the output that it would have provided if the request was
+actually executed.
+
+To provide this option, the method **should** include a `bool validate_only`
+field in the request message:
+
+```proto
+message ReviewBookRequest {
+  string name = 1 [(google.api.resource_reference) = {
+    type: "library-example.googleapis.com/Book"
+  }];
+  int32 rating = 2;
+  string comment = 3;
+
+  // If set, validate the request and preview the review, but do not actually
+  // post it.
+  bool validate_only = 4;
+}
+```
+
+The API **must** perform permission checks and any other validation that would
```

Comment:
Does this include existence checks? I think there are definitely contextual and non-contextual validation checks that can be performed - once we get into permissions and existence (e.g. yes, you can create this resource because it doesn't exist yet) I get nervous. On the other hand, I can see it being useful, particularly if it's "read existence" checks, e.g. "I want to refer to resources X and Y, so they ought to exist."

I'm slightly uncomfortable about this, but think it's *probably* the right call.

Reply:
I guess my rationale here is that the alternative seems wrong: if we _do not_ do the existence check, then the "validate_only" request says, "Your request is valid!" -- and then the actual request fails with ALREADY_EXISTS.

I get that the system state could change between the requests also, so `validate_only` is not foolproof. But it does seem like it should accurately represent validity at the time.

Reply:
Mmm. I think both "validate that this is a reasonable request" and "validate that this request would succeed now" are potentially useful, but yes, let's go with the latter for now.
(The former would be "An RPC with the same request mustn't fail with an INVALID_REQUEST error" (or whatever the status code is).)

Reply:
`INVALID_ARGUMENT`, but yes. Agree that both have value, and agree that the word "valid" technically means the former, but most users probably expect the latter.

Regardless, we are in concurrence.